File: python/RaspberryPi/phprobe.py
import math
import struct
import time
import smbus
global i2c

PH_PROBE = 0x3d

PH_MEASURE_PH = 80
PH_MEASURE_TEMP = 40
PH_CALIBRATE_SINGLE = 20
PH_CALIBRATE_LOW = 10
PH_CALIBRATE_HIGH = 8
PH_I2C = 1

PH_VERSION_REGISTER = 0
PH_PH_REGISTER = 1
PH_TEMP_REGISTER = 5
PH_CALIBRATE_SINGLE_REGISTER = 9
PH_CALIBRATE_REFHIGH_REGISTER = 13
PH_CALIBRATE_REFLOW_REGISTER = 17
PH_CALIBRATE_READHIGH_REGISTER = 21
PH_CALIBRATE_READLOW_REGISTER = 25
PH_SOLUTION_REGISTER = 29
PH_MV_REGISTER = 33
PH_CONFIG_REGISTER = 37
PH_TASK_REGISTER = 38

PH_TEMP_MEASURE_TIME = 750
PH_PH_MEASURE_TIME = 1100

PH_DUALPOINT_CONFIG_BIT = 0
PH_TEMP_COMPENSATION_CONFIG_BIT = 1


class phprobe(object):
    pH = 0
    pOH = 0
    mV = 0
    tempC = 0
    tempF = 0
    address = PH_PROBE

    def __init__(self, i2c_bus, **kwargs):
        global i2c
        i2c = smbus.SMBus(i2c_bus)

    def measureTemp(self):
        self._send_command(PH_MEASURE_TEMP)
        time.sleep(PH_TEMP_MEASURE_TIME / 1000.0)
        self.tempC = self._read_register(PH_TEMP_REGISTER)
        self.tempF = ((self.tempC * 9) / 5) + 32
        return self.tempC

    def measurepH(self):
        if self.usingTemperatureCompensation() is True:
            self.measureTemp()
        self._send_command(PH_MEASURE_PH)
        time.sleep(PH_PH_MEASURE_TIME / 1000.0)
        self.pH = self._read_register(PH_PH_REGISTER)
        self.mV = self._read_register(PH_MV_REGISTER) * 1000
        self.pH = round(self.pH * 100.0) / 100.0
        self.mV = round(self.mV * 100.0) / 100.0
        self.pOH = abs(self.pH - 14)

        if self.pH <= 0.0 or self.pH >= 14.0:
            self.pH = -1
            self.pOH = -1

        if math.isinf(self.pH):
            self.pH = -1
            self.pOH = -1

        if math.isnan(self.pH):
            self.pH = -1
            self.pOH = -1

        return self.pH

    def calibrateSingle(self, solutionpH):
        dualpoint = self.usingDualPoint()

        self.useDualPoint(0)
        self._write_register(PH_SOLUTION_REGISTER, solutionpH)
        self._send_command(PH_CALIBRATE_SINGLE)
        time.sleep(PH_PH_MEASURE_TIME / 1000.0)
        self.useDualPoint(dualpoint)

    def calibrateProbeLow(self, solutionpH):
        dualpoint = self.usingDualPoint()

        self.useDualPoint(0)
        self._write_register(PH_SOLUTION_REGISTER, solutionpH)
        self._send_command(PH_CALIBRATE_LOW)
        time.sleep(PH_PH_MEASURE_TIME / 1000.0)
        self.useDualPoint(dualpoint)

    def calibrateProbeHigh(self, solutionpH):
        dualpoint = self.usingDualPoint()

        self.useDualPoint(0)
        self._write_register(PH_SOLUTION_REGISTER, solutionpH)
        self._send_command(PH_CALIBRATE_HIGH)
        time.sleep(PH_PH_MEASURE_TIME / 1000.0)
        self.useDualPoint(dualpoint)

    def getVersion(self):
        return self._read_byte(PH_VERSION_REGISTER)

    def getCalibrateOffset(self):
        return self._read_register(PH_CALIBRATE_SINGLE_REGISTER)

    def getCalibrateHigh(self):
        return self._read_register(PH_CALIBRATE_REFHIGH_REGISTER)

    def getCalibrateLow(self):
        return self._read_register(PH_CALIBRATE_REFLOW_REGISTER)

    def getCalibrateHighReading(self):
        return self._read_register(PH_CALIBRATE_READHIGH_REGISTER)

    def getCalibrateLowReading(self):
        return self._read_register(PH_CALIBRATE_READLOW_REGISTER)

    def reset(self):
        n = float('nan')
        self._write_register(PH_CALIBRATE_SINGLE_REGISTER,   n)
        self._write_register(PH_CALIBRATE_REFHIGH_REGISTER,  n)
        self._write_register(PH_CALIBRATE_REFLOW_REGISTER,   n)
        self._write_register(PH_CALIBRATE_READHIGH_REGISTER, n)
        self._write_register(PH_CALIBRATE_READLOW_REGISTER,  n)

    def setDualPointCalibration(self, refLow, refHigh, readLow, readHigh):
        self._write_register(PH_CALIBRATE_REFLOW_REGISTER,   refLow)
        self._write_register(PH_CALIBRATE_REFHIGH_REGISTER,  refHigh)
        self._write_register(PH_CALIBRATE_READLOW_REGISTER,  readLow)
        self._write_register(PH_CALIBRATE_READHIGH_REGISTER, readHigh)

    def setI2CAddress(self, i2cAddress):
        self._write_byte(PH_SOLUTION_REGISTER, int(i2cAddress))
        self._send_command(PH_I2C)
        self.address = int(i2cAddress)

    def useTemperatureCompensation(self, b):
        retval = self._read_byte(PH_CONFIG_REGISTER)

        retval = self._bit_set(retval, PH_TEMP_COMPENSATION_CONFIG_BIT, b)
        self._write_byte(PH_CONFIG_REGISTER, retval)

    def useDualPoint(self, b):
        retval = self._read_byte(PH_CONFIG_REGISTER)

        retval = self._bit_set(retval, PH_DUALPOINT_CONFIG_BIT, b)
        self._write_byte(PH_CONFIG_REGISTER, retval)

    def usingTemperatureCompensation(self):
        retval = self._read_byte(PH_CONFIG_REGISTER)
        return (retval >> 1) & 0x01

    def usingDualPoint(self):
        retval = self._read_byte(PH_CONFIG_REGISTER)
        return (retval >> 0) & 0x01

    def _bit_set(self, v, index, x):
        mask = 1 << index
        v &= ~mask
        if x:
            v |= mask
        return v

    def _change_register(self, r):
        global i2c
        i2c.write_byte(self.address, r)
        time.sleep(10 / 1000.0)

    def _send_command(self, command):
        global i2c
        i2c.write_byte_data(self.address, PH_TASK_REGISTER, command)
        time.sleep(10 / 1000.0)

    def _write_register(self, reg, f):
        global i2c
        n = self.round_total_digits(f)
        fd = bytearray(struct.pack("f", n))
        data = [0, 0, 0, 0]
        data[0] = fd[0]
        data[1] = fd[1]
        data[2] = fd[2]
        data[3] = fd[3]
        self._change_register(reg)
        i2c.write_i2c_block_data(self.address, reg, data)
        time.sleep(10 / 1000.0)

    def _read_register(self, reg):
        global i2c
        data = [0, 0, 0, 0]
        self._change_register(reg)
        data[0] = i2c.read_byte(self.address)
        data[1] = i2c.read_byte(self.address)
        data[2] = i2c.read_byte(self.address)
        data[3] = i2c.read_byte(self.address)
        ba = bytearray(data)
        f = struct.unpack('f', ba)[0]
        return self.round_total_digits(f)

    def _write_byte(self, reg, val):
        global i2c
        i2c.write_byte_data(self.address, reg, val)
        time.sleep(10 / 1000.0)

    def _read_byte(self, reg):
        global i2c
        self._change_register(reg)
        time.sleep(10 / 1000.0)
        return i2c.read_byte(self.address)

    def magnitude(self, x):
        if math.isnan(x):
            return 0
        return 0 if x == 0 else int(math.floor(math.log10(abs(x)))) + 1

    def round_total_digits(self, x, digits=7):
        return round(x, digits - self.magnitude(x))

File: javascript/pHProbe.js
var C = {
  PH_PROBE: 0x3d,
  PH_MEASURE_PH: 80,
  PH_MEASURE_TEMP: 40,
  PH_CALIBRATE_SINGLE: 20,
  PH_CALIBRATE_LOW: 10,
  PH_CALIBRATE_HIGH: 8,
  PH_I2C: 1,
  PH_VERSION_REGISTER: 0,
  PH_PH_REGISTER: 1,
  PH_TEMP_REGISTER: 5,
  PH_CALIBRATE_SINGLE_REGISTER: 9,
  PH_CALIBRATE_REFHIGH_REGISTER: 13,
  PH_CALIBRATE_REFLOW_REGISTER: 17,
  PH_CALIBRATE_READHIGH_REGISTER: 21,
  PH_CALIBRATE_READLOW_REGISTER: 25,
  PH_SOLUTION_REGISTER: 29,
  PH_MV_REGISTER: 33,
  PH_CONFIG_REGISTER: 37,
  PH_TASK_REGISTER: 38,
  PH_TEMP_MEASURE_TIME: 750,
  PH_PH_MEASURE_TIME: 1100,
  PH_DUALPOINT_CONFIG_BIT: 0,
  PH_TEMP_COMPENSATION_CONFIG_BIT: 1
};

function pHProbe(_i2c, _address) {
  this.i2c = _i2c;
  this.address = _address;
}

pHProbe.prototype._send_command = function(cmd) {
  this.i2c.writeTo(this.address, [C.PH_TASK_REGISTER, cmd]);
  for (f = 0; f < 10000; f++) {}
};

pHProbe.prototype._read_byte = function(reg) {
  _change_register(reg);
  return this.i2c.readFrom(this.address, 1)[0];
  for (f = 0; f < 10000; f++) {}
};

pHProbe.prototype._write_byte = function(reg, val) {
  this.i2c.writeTo(this.address, [reg, val]);
  for (f = 0; f < 10000; f++) {}
};

pHProbe.prototype._change_register = function(reg) {
  this.i2c.writeTo(this.address, reg);
  for (f = 0; f < 10000; f++) {}
};

pHProbe.prototype._read_register = function(reg) {
  var data = new Uint8Array(4);

  this._change_register(reg);
  data[0] = this.i2c.readFrom(this.address, 1)[0];
  for (f = 0; f < 10000; f++) {}
  data[1] = this.i2c.readFrom(this.address, 1)[0];
  for (f = 0; f < 10000; f++) {}
  data[2] = this.i2c.readFrom(this.address, 1)[0];
  for (f = 0; f < 10000; f++) {}
  data[3] = this.i2c.readFrom(this.address, 1)[0];
  for (f = 0; f < 10000; f++) {}

  return new Float32Array(data.buffer)[0];
};

pHProbe.prototype._write_register = function(reg, val) {
  var data = new Float32Array(1);
  data[0] = val;

  this.i2c.writeTo(this.address, [reg, data.buffer[0], data.buffer[1], data.buffer[2], data.buffer[3]]);
  for (f = 0; f < 10000; f++) {}
};

exports.connect = function(_i2c, _address) {
  return new pHProbe(_i2c, _address);
};

pHProbe.prototype.measureTemp = function(callback) {
  var tempC;
  var tempF;
  var self = this;

  this._send_command(C.PH_MEASURE_TEMP);
  var i = setInterval(function() {
    tempC = self._read_register(C.PH_TEMP_REGISTER);
    tempF = ((tempC * 9) / 5) + 32;
    callback({
      tempC: tempC,
      tempF: tempF
    });
    clearInterval(i);
  }, C.PH_TEMP_MEASURE_TIME);
};

pHProbe.prototype.measurepH = function(callback) {
  var pH;
  var mV;
  var pOH;
  var self = this;

  if (this.useTemperatureCompensation == true) {
    this.measureTemp();
  }
  this._send_command(C.PH_MEASURE_PH);
  var i = setInterval(function() {
    pH = self._read_register(C.PH_PH_REGISTER);
    mV = self._read_register(C.PH_MV_REGISTER) * 1000;
    pH = Math.round(pH * 100.0) / 100.0;
    mV = Math.round(mV * 100.0) / 100.0;
    pOH = Math.abs(pH - 14);
    callback({
      pH: pH,
      mV: mV,
      pOH: pOH
    });
    clearInterval(i);
  }, C.PH_PH_MEASURE_TIME);
};

pHProbe.prototype.calibrateSingle = function(solutionpH, callback) {
  var dualpoint = this.usingDualPoint();
  var self = this;

  this.useDualPoint(false);
  this._write_register(C.PH_SOLUTION_REGISTER, solutionpH);
  this._send_command(C.PH_CALIBRATE_SINGLE);
  this.useDualPoint(dualpoint);
  var i = setInterval(function() {
    callback({
      offset: self._read_register(C.PH_CALIBRATE_SINGLE_REGISTER)
    });
    clearInterval(i);
  }, C.PH_PH_MEASURE_TIME);
};

pHProbe.prototype.calibrateLow = function(solutionpH) {
  var dualpoint = this.usingDualPoint();
  var self = this;

  this.useDualPoint(false);
  this._write_register(C.PH_SOLUTION_REGISTER, solutionpH);
  this._send_command(C.PH_CALIBRATE_LOW);
  this.useDualPoint(dualpoint);
};

pHProbe.prototype.calibrateHigh = function(solutionpH) {
  var dualpoint = this.usingDualPoint();
  var self = this;

  this.useDualPoint(false);
  this._write_register(C.PH_SOLUTION_REGISTER, solutionpH);
  this._send_command(C.PH_CALIBRATE_HIGH);
  //time.sleep(PH_PH_MEASURE_TIME / 1000.0)
  this.useDualPoint(dualpoint);
};

pHProbe.prototype.getVersion = function() {
  return this._read_byte(C.PH_VERSION_REGISTER);
};

pHProbe.prototype.getCalibrateOffset = function() {
  return Math.round(this._read_register(C.PH_CALIBRATE_SINGLE_REGISTER) * 100.0) / 100.0;
};

pHProbe.prototype.getCalibrateHigh = function() {
  return Math.round(this._read_register(C.PH_CALIBRATE_REFHIGH_REGISTER) * 100.0) / 100.0;;
};

pHProbe.prototype.getCalibrateLow = function() {
  return Math.round(this._read_register(C.PH_CALIBRATE_REFLOW_REGISTER) * 100.0) / 100.0;;
};

pHProbe.prototype.getCalibrateHighReading = function() {
  return Math.round(this._read_register(C.PH_CALIBRATE_READHIGH_REGISTER) * 100.0) / 100.0;;
};

pHProbe.prototype.getCalibrateLowReading = function() {
  return Math.round(this._read_register(C.PH_CALIBRATE_READLOW_REGISTER) * 100.0) / 100.0;;
};

pHProbe.prototype.reset = function() {
  this._write_register(C.PH_CALIBRATE_SINGLE_REGISTER, 1);
  this._write_register(C.PH_CALIBRATE_REFHIGH_REGISTER, 1);
  this._write_register(C.PH_CALIBRATE_REFLOW_REGISTER, 1);
  this._write_register(C.PH_CALIBRATE_READHIGH_REGISTER, 1);
  this._write_register(C.PH_CALIBRATE_READLOW_REGISTER, 1);
};

pHProbe.prototype.setDualPointCalibration = function(refLow, refHigh, readLow, readHigh) {
  this._write_register(C.PH_CALIBRATE_REFLOW_REGISTER, refLow);
  this._write_register(C.PH_CALIBRATE_REFHIGH_REGISTER, refHigh);
  this._write_register(C.PH_CALIBRATE_READLOW_REGISTER, readLow);
  this._write_register(C.PH_CALIBRATE_READHIGH_REGISTER, readHigh);
};

pHProbe.prototype.setI2CAddress = function(i2cAddress) {
  this._write_byte(C.PH_SOLUTION_REGISTER, i2cAddress);
  this._send_command(C.PH_I2C);
  this.address = i2cAddress;
};

pHProbe.prototype.useTemperatureCompensation = function(b) {
  var retval = this._read_byte(C.PH_CONFIG_REGISTER);

  retval = this._bit_set(retval, C.PH_TEMP_COMPENSATION_CONFIG_BIT, b);
  this._write_byte(C.PH_CONFIG_REGISTER, retval);
};

pHProbe.prototype.useDualPoint = function(b) {
  var retval = this._read_byte(C.PH_CONFIG_REGISTER);

  retval = this._bit_set(retval, C.PH_DUALPOINT_CONFIG_BIT, b);
  this._write_byte(C.PH_CONFIG_REGISTER, retval);
};

pHProbe.prototype.usingTemperatureCompensation = function() {
  var retval = this._read_byte(C.PH_CONFIG_REGISTER);
  return (retval >> 1) & 0x01;
};

pHProbe.prototype.usingDualPoint = function() {
  var retval = this._read_byte(C.PH_CONFIG_REGISTER);
  return (retval >> 0) & 0x01;
};

pHProbe.prototype._bit_set = function(v, index, x) {
  var mask = 1 << index;
  v &= ~mask;
  if (x) {
    v |= mask;
  }
  return v;
};